### PR TITLE
UrlDispatcher fixes: #2598, #2645, #2601, #2614, #2567

### LINKED ISSUE #2598: Seo friendly urls - bugs
## Current behavior

1. If I disable seo URL I have different category link in the menu and in breadcrumbs.
In breadcrumbs, I have the correct link with 'c' prefix
In menu I don't have it "c" prefix -> URL is still generated base on 'full_path'

2. If I go directly to subcategory page
https://next.storefrontcloud.io/men/tops-men/jackets-men/jackets-14
I don't see full menu -> maybe it is connected to the first issue.

## Expected behavior

1. Regardless of new config setting: `seo.useUrlDispatcher`, we should always get the same URL for the category in the menu and in breadcrumbs.

## Steps to reproduce the issue

1. set `seo.useUrlDispatcher` to false I check category URL in the menu and in breadcrumbs

## Can you handle fixing this bug by yourself?

NO

### PR DESCRIPTION
**Pull request by pkarw**
### Related issues

closes #2598, #2645, #2601, #2614, #2567 

### Upgrade Notes and Changelog

- [x] No upgrade steps required (100% backward compatibility)
- [ ] I've updated the [Upgrade notes](https://github.com/DivanteLtd/vue-storefront/blob/develop/doc/Upgrade%20notes.md) and [Changelog](https://github.com/DivanteLtd/vue-storefront/blob/develop/CHANGELOG.md) on how to port existing VS sites with this new feature

### Contribution and currently important rules acceptance

- [x] I read and followed [contribution rules](https://github.com/DivanteLtd/vue-storefront/blob/master/CONTRIBUTING.md)
- [x] I read the [TypeScript Action Plan](https://github.com/DivanteLtd/vue-storefront/blob/master/doc/TypeScript%20Action%20Plan.md) and adjusted my PR according to it
- [x] I read about [Vue Storefront Modules](https://github.com/DivanteLtd/vue-storefront/blob/master/doc/api-modules/about-modules.md) and I am aware that every new feature should be a module

### CODE REVIEW

File: core/modules/cart/store/actions.ts [hunk 1]
```diff
@@ -335,7 +335,7 @@ const actions: ActionTree<CartState, RootState> = {
     commit(types.CART_UPD_ITEM_PROPS, { product })
   },
   getPaymentMethods (context) {
-    if (config.cart.synchronize_totals && (typeof navigator !== 'undefined' ? navigator.onLine : true)) {
+    if (config.cart.synchronize_totals && (typeof navigator !== 'undefined' ? navigator.onLine : true) && context.rootState.cart.cartServerToken) {
```
Comment (patzick) [inline core/modules/cart/store/actions.ts]:
```suggestion
    if (config.cart.synchronize_totals && onlineHelper.isOnline && context.rootState.cart.cartServerToken) {
```

File: core/modules/cart/store/actions.ts [hunk 2]
```diff
@@ -365,7 +365,7 @@ const actions: ActionTree<CartState, RootState> = {
   },
   getShippingMethods (context, address) {
     return new Promise((resolve, reject) => {
-      if (config.cart.synchronize_totals && (typeof navigator !== 'undefined' ? navigator.onLine : true)) {
+      if (config.cart.synchronize_totals && (typeof navigator !== 'undefined' ? navigator.onLine : true) && context.rootState.cart.cartServerToken) {
```
Comment (patzick) [inline core/modules/cart/store/actions.ts]:
```suggestion
      if (config.cart.synchronize_totals && onlineHelper.isOnline && context.rootState.cart.cartServerToken) {
```

File: core/modules/cart/store/actions.ts [hunk 3]
```diff
@@ -391,7 +391,7 @@ const actions: ActionTree<CartState, RootState> = {
   refreshTotals (context, methodsData) {
     return new Promise((resolve, reject) => {
       const storeView = currentStoreView()
-      if (config.cart.synchronize_totals) {
+      if (config.cart.synchronize_totals && (typeof navigator !== 'undefined' ? navigator.onLine : true) && context.rootState.cart.cartServerToken) {
```
Comment (patzick) [inline core/modules/cart/store/actions.ts]:
```suggestion
      if (config.cart.synchronize_totals && onlineHelper.isOnline && context.rootState.cart.cartServerToken) {
```

File: core/pages/PageNotFound.js [hunk 4]
```diff
@@ -14,22 +13,20 @@ export default {
...
-          }
-        })
+      store.dispatch('product/list', {
+        query: ourBestsellersQuery,
+        size: 8,
+        sort: 'created_at:desc'
+      }).then(res => {
+        if (res) {
+          store.state.homepage.bestsellers = res.items
+          EventBus.$emitFilter('pagenotfound-after-load', { store: store, route: route }).then(results => {
```
Comment (patzick) [inline core/pages/PageNotFound.js]: Where is this used? Haven't found another `pagenotfound-after-load` string.

File: core/lib/multistore.ts [hunk 5]
```diff
@@ -113,7 +113,7 @@ export function adjustMultistoreApiUrl (url: string) : string {
 }
 
 export function localizedRoute (routeObj: LocalizedRoute | string | RouteConfig | RawLocation, storeCode: string): any {
-  if (typeof routeObj === 'object' && routeObj && (<LocalizedRoute>routeObj).fullPath) return localizedDispatcherRoute(<LocalizedRoute>Object.assign({}, routeObj, { params: null }), storeCode)
+  if (typeof routeObj === 'object' && routeObj && (<LocalizedRoute>routeObj).fullPath && rootStore.state.config.seo.useUrlDispatcher) return localizedDispatcherRoute(<LocalizedRoute>Object.assign({}, routeObj, { params: null }), storeCode)
```
Comment (filrak) [inline core/lib/multistore.ts]: `<LocalizedRoute>routeObj` -type is set 1 line above
`typeof routeObj === 'object' && routeObj` - isn't 1st check redundant?

File: core/lib/sync/index.ts [hunk 6]
```diff
@@ -60,7 +60,7 @@ function execute (task) { // not offline task
           if (!currentCartId && rootStore.state.cart.cartServerToken) { // this is workaround; sometimes after page is loaded indexedb returns null despite the cart token is properly set
             currentCartId = rootStore.state.cart.cartServerToken
           }
-          if (!currentToken && rootStore.state.user.cartServerToken) { // this is workaround; sometimes after page is loaded indexedb returns null despite the cart token is properly set
+          if (!currentToken && rootStore.state.user.token) { // this is workaround; sometimes after page is loaded indexedb returns null despite the cart token is properly set
```
Comment (filrak) [inline core/lib/sync/index.ts]: can we make it a proper getter?

File: core/modules/cart/store/actions.ts [hunk 7]
```diff
@@ -59,7 +59,7 @@ const actions: ActionTree<CartState, RootState> = {
     context.commit(types.CART_SAVE)
   },
   serverPull (context, { forceClientState = false, dryRun = false }) { // pull current cart FROM the server
-    if (config.cart.synchronize && !isServer) {
+    if (config.cart.synchronize && !isServer && (typeof navigator !== 'undefined' ? navigator.onLine : true) && context.state.cartServerToken) {
```
Comment (filrak) [inline core/modules/cart/store/actions.ts]: we have `isOnline` helper for this

File: core/modules/cart/store/actions.ts [hunk 8]
```diff
@@ -94,7 +94,7 @@ const actions: ActionTree<CartState, RootState> = {
     }
   },
   serverTotals (context, { forceClientState = false }) { // pull current cart FROM the server
-    if (config.cart.synchronize_totals  && !isServer) {
+    if (config.cart.synchronize_totals  && !isServer && (typeof navigator !== 'undefined' ? navigator.onLine : true) && context.state.cartServerToken) {
```
Comment (filrak) [inline core/modules/cart/store/actions.ts]: there is `isOnline` helper

**Comment by patzick** [comment]
Closing in favor of #2652

**Pull request closed without merging**